src/internal/common/readers/serialization_version.go

```diff
@@ -0,0 +1,181 @@
+//	     +----------+----+---------+--------+-------+
+type persistedSerializationVersion = uint32
+
+// SerializationVersion is the in-memory size of the version number that gets
+// added to the persisted serialization version.
+//
+// Right now it's only a uint16 but we can expand it to be larger so long as the
+// expanded size doesn't clash with the flags in the high-order bits.
+type SerializationVersion uint16
+
+// DefaultSerializationVersion is the current (default) version number for all
+// services. As services evolve their storage format they should begin tracking
+// their own version numbers separate from other services.
+const DefaultSerializationVersion SerializationVersion = 1
+
+var (
+	versionFormatSize                               = int(unsafe.Sizeof(persistedSerializationVersion(0)))
+	delInFlightMask   persistedSerializationVersion = 1 << ((versionFormatSize * 8) - 1)
+)
+
+// SerializationFormat is a struct describing serialization format versions and
+// flags to add for this item.
+type SerializationFormat struct {
+	Version     SerializationVersion
+	DelInFlight bool
+}
+
+func NewVersionedBackupReader(
+	format SerializationFormat,
+	baseReaders ...io.ReadCloser,
+) (io.ReadCloser, error) {
+	if format.DelInFlight && len(baseReaders) > 0 {
+		// This is a conservative check, but we can always loosen it later on if
+		// needed. At the moment we really don't expect any data if the item was
+		// deleted.
+		return nil, clues.New("item marked deleted but has reader(s)")
+	}
+
+	formattedVersion := persistedSerializationVersion(format.Version)
+	if format.DelInFlight {
+		formattedVersion |= delInFlightMask
+	}
+
+	formattedBuf := make([]byte, versionFormatSize)
+	binary.BigEndian.PutUint32(formattedBuf, formattedVersion)
+
+	versionReader := io.NopCloser(bytes.NewReader(formattedBuf))
+
+	// Need to add readers individually because types differ.
+	allReaders := make([]io.Reader, 0, len(baseReaders)+1)
+	allReaders = append(allReaders, versionReader)
+
+	for _, r := range baseReaders {
+		allReaders = append(allReaders, r)
+	}
+
+	res := &versionedBackupReader{
+		baseReaders: append([]io.ReadCloser{versionReader}, baseReaders...),
+		combined:    io.MultiReader(allReaders...),
+	}
+
+	return res, nil
+}
+
+type versionedBackupReader struct {
+	// baseReaders is a reference to the original readers so we can close them.
+	baseReaders []io.ReadCloser
+	// combined is the reader that will return all data.
+	combined io.Reader
+}
+
+func (vbr *versionedBackupReader) Read(p []byte) (int, error) {
+	if vbr.combined == nil {
+		return 0, os.ErrClosed
+	}
+
+	n, err := vbr.combined.Read(p)
+	if err == io.EOF {
+		// Golang doesn't allow wrapping of EOF. If we wrap it other things start
+		// thinking it's an actual error.
+		return n, err
+	}
+
+	return n, clues.Stack(err).OrNil()
+}
+
+func (vbr *versionedBackupReader) Close() error {
+	if vbr.combined == nil {
+		return nil
+	}
+
+	vbr.combined = nil
+
+	var errs *clues.Err
```

Comment:
hmm, this feels more like a case for fault recoverables than errs.  Create a fault bus here, add recoverable errors throughout.  If recoverable > 0, return a clean sentinel while logging each error case. 

My concern with using clues stacking here is that its format implies that deeper stacks occurred within context of earlier ones, not distinct from them (wtb `clues.Tree(err, errs...)`?  No, just kidding. ...maybe).  Which could make debugging a bit odd.  I think what you really mean to communicate is that 1/ at least one closer failed (return value), and 2/ one or more errors occurred (logs).  Easier with fault, harder with clues.

Reply:
unfortunately using fault requires a context for logging. I don't really want to have to keep a reference to that here (and pass it in in all the places this will be used). The current option will work well enough even if the error output is a tad odd. I think having the `reader_index` clue in there will help readers figure out what's going on when viewed in conjunction with this code